File: phoenix-api/lib/app_web/controllers/user_controller.ex
defmodule AppWeb.UserController do
  use AppWeb, :controller

  alias App.Accounts
  alias AppWeb.UserParams

  # GET /users
  def index(conn, params) do
    users = Accounts.list_users(params)
    json(conn, users)
  end

  # GET /users/:id
  def show(conn, %{"id" => id}) do
    id
    |> String.trim()
    |> Ecto.UUID.cast()
    |> case do
      {:ok, uuid} ->
        case Accounts.get_user(uuid) do
          nil -> send_resp(conn, 404, "")
          user -> json(conn, user)
        end

      :error ->
        send_resp(conn, 400, "Invalid UUID")
    end
  end

  # POST /users
  def create(conn, params) do
    case UserParams.cast_and_validate(params) do
      %Ecto.Changeset{valid?: true} = changeset ->
        case Accounts.create_user(changeset.changes) do
          {:ok, user} -> json(conn, user)
          {:error, cs} -> conn |> put_status(422) |> json(%{errors: cs})
        end

      changeset ->
        conn |> put_status(400) |> json(%{errors: changeset})
    end
  end

  # PUT /users/:id
  def update(conn, %{"id" => id} = params) do
    id
    |> String.trim()
    |> Ecto.UUID.cast()
    |> case do
      {:ok, uuid} ->
        case Accounts.get_user(uuid) do
          nil ->
            send_resp(conn, 404, "")

          _user ->
            case UserParams.cast_and_validate(params) do
              %Ecto.Changeset{valid?: true} = changeset ->
                case Accounts.update_user(uuid, changeset.changes) do
                  {:ok, user} -> json(conn, user)
                  {:error, cs} -> conn |> put_status(422) |> json(%{errors: cs})
                end

              changeset ->
                conn |> put_status(400) |> json(%{errors: changeset})
            end
        end

      :error ->
        send_resp(conn, 400, "Invalid UUID")
    end
  end

  # DELETE /users/:id
  def delete(conn, %{"id" => id}) do
    id
    |> String.trim()
    |> Ecto.UUID.cast()
    |> case do
      {:ok, uuid} ->
        case Accounts.get_user(uuid) do
          nil ->
            send_resp(conn, 404, "")

          _user ->
            Accounts.delete_user(uuid)
            send_resp(conn, 204, "")
        end

      :error ->
        send_resp(conn, 400, "Invalid UUID")
    end
  end
end

File: symfony-app/src/UserInterface/Web/UserController.php
<?php

declare(strict_types=1);

namespace App\UserInterface\Web;

use App\Domain\User\Port\UserApiClientInterface;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\JsonResponse;
use Symfony\Component\Routing\Annotation\Route;
use Symfony\Component\HttpFoundation\Request;

#[Route('/api/users')]
class UserController extends AbstractController
{
    private UserApiClientInterface $userApi;

    public function __construct(UserApiClientInterface $userApi)
    {
        $this->userApi = $userApi;
    }

    #[Route('', name: 'list_users', methods: ['GET'])]
    public function list(Request $request): JsonResponse
    {
        $filters = $request->query->all();
        
        return $this->json($this->userApi->list($filters));
    }

    #[Route('/{id}', name: 'get_user', methods: ['GET'])]
    public function get(string $id): JsonResponse
    {
        $user = $this->userApi->get($id);
        if (!$user) return $this->json(['error' => 'Not found'], 404);

        return $this->json($user);
    }

    #[Route('', name: 'create_user', methods: ['POST'])]
    public function create(Request $request): JsonResponse
    {
        $data = json_decode($request->getContent(), true);

        return $this->json($this->userApi->create($data));
    }

    #[Route('/{id}', name: 'update_user', methods: ['PUT'])]
    public function update(string $id, Request $request): JsonResponse
    {
        $data = json_decode($request->getContent(), true);

        return $this->json($this->userApi->update($id, $data));
    }

    #[Route('/{id}', name: 'delete_user', methods: ['DELETE'])]
    public function delete(string $id): JsonResponse
    {
        $deleted = $this->userApi->delete($id);

        return $this->json(null, $deleted ? 204 : 404);
    }

    #[Route('/import', name: 'import_users', methods: ['POST'])]
    public function import(Request $request): JsonResponse
    {
        $data = json_decode($request->getContent(), true);

        return $this->json($this->userApi->import($data));
    }
}